plugins/runtime/runtime.go

```diff
@@ -0,0 +1,313 @@
+
+import (
+	"context"
+	"errors"
+	goruntime "runtime"
+	"sync"
+	"time"
+
+	"go.opentelemetry.io/otel/api/metric"
+	"go.opentelemetry.io/otel/api/unit"
+)
+
+// Runtime reports the work-in-progress conventional runtime metrics specified by OpenTelemetry
+type Runtime struct {
+	mu       sync.RWMutex
+	meter    metric.Meter
+	interval time.Duration
+	done     chan bool
+
+	metrics struct {
+		goCgoCalls metric.Int64Counter
+		goLookups  metric.Int64Counter
+		goGcCount  metric.Int64Counter
+		gcPauseNs  metric.Int64Measure
+	}
+
+	memStats    goruntime.MemStats
+	numCgoCalls int64
+}
+
+// New returns Runtime, a structure for reporting Go runtime metrics
+// interval is used to define how often to invoke Go runtime.ReadMemStats() to obtain metric data. It should be noted
+// this package invokes a stop-the-world function on this interval. The interval should not be set arbitrarily small
+// without accepting the performance overhead.
+// TODO this interval may be removed in favor of otel SDK control after batch observers land
+func New(meter metric.Meter, interval time.Duration) *Runtime {
+	r := &Runtime{
+		meter:    meter,
+		interval: interval,
+		done:     make(chan bool),
+	}
+
+	return r
+}
+
+// Start begins regular background polling of Go runtime metrics and will return an error if any issues are encountered
+func (r *Runtime) Start() error {
+	if r.interval <= 0 {
+		return errors.New("non-positive interval for runtime.New")
+	}
+
+	err := r.register()
+	if err != nil {
+		return err
+	}
+
+	go r.ticker()
+
+	return nil
+}
+
+// Stop terminates the regular background polling of Go runtime metrics
+func (r *Runtime) Stop() {
+	r.done <- true
+}
+
+func (r *Runtime) ticker() {
+	ticker := time.NewTicker(r.interval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-r.done:
+			return
+		case <-ticker.C:
+			ctx := context.Background()
+			r.collect(ctx)
+		}
+	}
+}
+
+func (r *Runtime) register() error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	var err error
+
+	t0 := time.Now()
+	_, err = r.meter.RegisterInt64Observer("runtime.uptime",
+		func(result metric.Int64ObserverResult) {
+			result.Observe(time.Since(t0).Milliseconds())
+		},
+		metric.WithUnit(unit.Milliseconds),
+		metric.WithDescription("Milliseconds since application was initialized"),
+	)
+	if err != nil {
+		return err
+	}
+
+	_, err = r.meter.RegisterInt64Observer("runtime.go.goroutines", func(result metric.Int64ObserverResult) {
+		result.Observe(int64(goruntime.NumGoroutine()))
+	}, metric.WithDescription("Number of goroutines that currently exist"))
+	if err != nil {
+		return err
+	}
+
+	r.metrics.goCgoCalls, err = r.meter.NewInt64Counter("runtime.go.cgo.calls",
+		metric.WithDescription("Number of cgo calls made by the current process"))
+	if err != nil {
+		return err
+	}
+
+	// poll now so that the first tick has a full delta
+	r.numCgoCalls = goruntime.NumCgoCall()
+	goruntime.ReadMemStats(&r.memStats)
```

Comment:
@quentinmit, reading the meeting nodes of the metrics sig, you mentioned that this ReadMemStats function (GC) "is no longer expensive" in Go?   Its still labeled as stop-the-world however. I'm curious to hear from folks who have more expertise than me in this area.

Reply:
I found this: https://go-review.googlesource.com/c/go/+/34937/
(Not sure..)

Reply:
Yeah, that's the change I was thinking of. Also, I believe [ReadGCStats](https://golang.org/pkg/runtime/debug/#ReadGCStats) does not STW at all.